backend/spvm.pm

```diff
@@ -95,4 +95,21 @@ sub stop_serial_grab {
     return;
 }
 
+sub power {
+    # parameters: reset, on, off
+    my ($self, $args) = @_;
+    my $action  = $args->{action};
+    my $lpar_id = get_required_var('NOVALINK_LPAR_ID');
+
+    if ($action eq 'reset') {
+        $self->run_cmd("pvmctl lpar restart -i id=${lpar_id}");
+    }
+    elsif ($action eq 'on') {
+        $self->run_cmd("pvmctl lpar power-on -i id=${lpar_id} --bootmode norm");
+    }
+    elsif ($action eq 'off') {
+        $self->run_cmd("pvmctl lpar power-off -i id=${lpar_id} --hard");
+    }
```

Comment:
I recommend to always have an else-branch, i.e.

```perl
else {
    die "Unknown action: '$action'";
}
```

but you might be better of with a function table:

```perl
%cmds = (
    reset => sub { $self->run_cmd("pvmctl lpar restart -i id=${lpar_id}") },
    on => sub { $self->run_cmd("pvmctl lpar power-on -i id=${lpar_id} --bootmode norm") },
    off => sub { $self->run_cmd("pvmctl lpar power-off -i id=${lpar_id} --hard") },
);
$cmds{$action}->();
```

no warranty for correct syntax, I just wrote "pseudo code" that might look like valid perl code but isn't ;)

Reply:
Having a function table make sense but you should still handle the case that the action is unknown then.

Reply:
That's at least implicitly covered with at least a perl warning trying to dereference an undefined hash value ;)

Reply:
@okurz  undefined value is handled now